File: src/main/java/at/phe/def/mapreduce/TuplePartitioner.java
package at.phe.def.mapreduce;

import at.phe.def.mapreduce.partitioner.IPartitioner;
import com.google.gson.JsonArray;

import java.util.ArrayList;
import java.util.List;

/**
 * Author: Philip Heimböck
 * Date: 23.04.16.
 */
public class TuplePartitioner {

    protected int numberReducers = 0;
    protected List<JsonArray> partitions = new ArrayList<>();
    protected IPartitioner partitioner;

    public TuplePartitioner(IPartitioner partitioner, int numberReducers) {
        this.partitioner = partitioner;
        this.numberReducers = numberReducers;

        for (int i = 0; i < numberReducers; i++) {
            partitions.add(new JsonArray());
        }
    }

    public void partition(JsonArray input) {
        // Partition the input
        input.forEach(tuple -> {
            // The tuple is another array
            JsonArray data = tuple.getAsJsonArray();

            // Choose always the same reducer for the same key
            int chosenReducer = partitioner.partition(data.get(0), numberReducers);
            partitions.get(chosenReducer).add(data);
        });
    }

    public List<JsonArray> getPartitions() {
        return partitions;
    }
}

File: src/main/dotnet/MapReduce/MapReduceBase/TuplePartitioner.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;

namespace MapReduceBase
{
	public class TuplePartitioner
	{
		public IPartitioner Partitioner 
		{
			get;
			set;
		}

		protected int NumberReducers {
			get;
			set;
		}

		protected List<JArray> partitions = new List<JArray>();

		public TuplePartitioner(IPartitioner partitioner, int numberReducers) {
			Partitioner = partitioner;
			NumberReducers = numberReducers;

			for (int i = 0; i < NumberReducers; i++) {
				partitions.Add(new JArray());
			}
		}

		public void partition(JArray input) {
			// Partition the input
			foreach (var tuple in input) {
				int chosenReducer = Partitioner.Partition (tuple [0], NumberReducers);
				partitions [chosenReducer].Add (tuple);
			}
		}

		public List<JArray> getPartitions() {
			return partitions;
		}
	}
}

